plugins/inputs/xtremio/xtremio.go

```diff
@@ -0,0 +1,324 @@
+package xtremio
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/common/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/tidwall/gjson"
+)
+
+type XtremIO struct {
+	Username   string   `toml:"username"`
+	Password   string   `toml:"password"`
+	URL        string   `toml:"url"`
+	Collectors []string `toml:"collectors"`
+	Cookie     *http.Cookie
+	tls.ClientConfig
+	Log telegraf.Logger
+}
+
+const sampleConfig = `
+  ## XtremIO Username
+  username = "user1" # required
+  ## XtremIO Password
+  password = "pass123" # required
+  ## XtremIO User Interface Endpoint
+  url = "https://xtremio.example.com/" # required
+  ## Metrics to collect from the XtremIO
+  collectors = ["bbus","clusters","ssds","volumes","xms"]
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use SSL but skip chain & host verification
+  # insecure_skip_verify = false
+`
+
+// Description will appear directly above the plugin definition in the config file
+func (xio *XtremIO) Description() string {
+	return `Gathers Metrics From a Dell EMC XtremIO Storage Array's V3 API`
+}
+
+// SampleConfig will populate the sample configuration portion of the plugin's configuration
+func (xio *XtremIO) SampleConfig() string {
+	return sampleConfig
+}
+
+func (xio *XtremIO) Init() error {
+	if xio.Username == "" {
+		return fmt.Errorf("Username cannot be empty")
+	}
+	if xio.Password == "" {
+		return fmt.Errorf("Password cannot be empty")
+	}
+	if xio.URL == "" {
+		return fmt.Errorf("URL cannot be empty")
+	}
+
+	return nil
+}
+
+func (xio *XtremIO) Gather(acc telegraf.Accumulator) error {
+	tlsCfg, err := xio.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+	http.DefaultTransport.(*http.Transport).TLSClientConfig = tlsCfg
+	err = xio.Authenticate()
+	if err != nil {
+		return err
+	}
+	if xio.Cookie == nil {
+		return fmt.Errorf("no authentication cookie set")
+	}
+
+	availableCollectors := []string{"bbus", "clusters", "ssds", "volumes", "xms"}
+	var wg sync.WaitGroup
+	for _, collector := range xio.Collectors {
+		wg.Add(1)
+		go func(collector string) {
+			defer wg.Done()
+			if !contains(availableCollectors, collector) {
+				acc.AddError(fmt.Errorf("Specified Collector Isn't Supported: " + collector))
+				return
+			}
+
+			resp, err := xio.Call(collector)
+			if err != nil {
+				acc.AddError(err)
+				return
+			}
+			var arr []gjson.Result
+			if collector == "xms" {
+				arr = gjson.Get(resp, "xmss").Array()
+			} else {
```

Comment:
Can you please normalize that name in `Init()` or accept `xmss` instead of `xms` to get rid of this exception?

Reply:
So after changing this bit, I realized why I had it there originally. It comes down to an inconsistency in the API. The endpoint is `xms` but the results returned are `xmss`. See example output below

```
{
    "xmss": [
        {
            "href": "https://10.0.0.10/api/json/v3/types/xms/xms_id_here", 
            "name": "xms"
        }
    ], 
    "links": [
        {
            "href": "https://10.0.0.10/api/json/v3/types/xms/", 
            "rel": "self"
        }
    ]
}
```